### PR TITLE
Fix darwin native leak

### PR DESCRIPTION
**Pull request by clementetb**
A reference cycle between `CPointerWrapper` and its cleaner was causing a leak were the GC was not reclaiming any native memory. 

This PR breaks the cycle by moving the `CPointerWrapper` and cleaner shared data into a new data class.

See [LeakingCleaner](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.native.ref/create-cleaner.html).

### CODE REVIEW

File: packages/cinterop/src/nativeDarwin/kotlin/io/realm/kotlin/internal/interop/RealmInterop.kt [hunk 1]
```diff
@@ -160,11 +159,22 @@ private fun <T : CPointed> checkedPointerResult(pointer: CPointer<T>?): CPointer
...
 }
 
-// FIXME API-INTERNAL Consider making NativePointer/CPointerWrapper generic to enforce typing
+/**
+ * Class with a pointer reference and its status. It breaks the reference cycle between CPointerWrapper
+ * and its GC cleaner, otherwise the cleaner would never be invoked.
+ *
+ * See leaking cleaner: https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.native.ref/create-cleaner.html
+ */
+data class ReleasablePointer(
```
Comment (cmelchior) [inline packages/cinterop/src/nativeDarwin/kotlin/io/realm/kotlin/internal/interop/RealmInterop.kt:168]: After reading those docs, I'm still not 100% sure why we need this class 🤔 

Couldn't we just move all the logic to the `release` method:

```
    val cleaner = if (managed) {
        createCleaner(this) { it.release() }
    } else null

    override fun release() {
        if (released.compareAndSet(expect = false, update = true)) {
            realm_release(_ptr)
        }
    }
```

I think it was me that added the old logic, and I don't remember if there was a problem not doing that in the first place 🙈

**Review by cmelchior — COMMENTED** [review]
Great find 🥇 I do have a question about the implementation though.

File: packages/cinterop/src/nativeDarwin/kotlin/io/realm/kotlin/internal/interop/RealmInterop.kt (hunk 1 above)
Comment (clementetb) [reply packages/cinterop/src/nativeDarwin/kotlin/io/realm/kotlin/internal/interop/RealmInterop.kt:168]: The original issue is that the cleaner lambda referenced a property of the object to clean, causing the object to never be garbage collected.

This class breaks the cycle, the cleaner lambda now points to it, allowing the GC to collect the original reference.

**Review by cmelchior — APPROVED** [review]
Great find. Just needs a changelog update, otherwise 👍

File: packages/cinterop/src/nativeDarwin/kotlin/io/realm/kotlin/internal/interop/RealmInterop.kt (hunk 1 above)
Comment (nhachicha) [reply packages/cinterop/src/nativeDarwin/kotlin/io/realm/kotlin/internal/interop/RealmInterop.kt:168]: I think in the original implementation we were referencing `released: AtomicBoolean = atomic(false)` inside the cleaner lambda (`released.compareAndSet(expect = false, update = true)`) which is equivalent to capturing `this` which will leak as per the doc example (`createCleaner(this)`).

**Review by nhachicha — APPROVED** [review]
Good catch 👍

**Pull request merged**